### PR TITLE
feat(get-cluster-info): reporting on eks clusters versions

### PR DESCRIPTION
**Pull request by jlapierre0istream**
This action gathers info and creates a markdown file that contains each EKS cluster in k8s management including name, region, EKS version and helm version.

### CODE REVIEW

**Review by abatilo — CHANGES_REQUESTED** [review]
What is this for?

**Comment by jlapierre0istream** [comment]
> What is this for?

@abatilo I added a description

**Review by marcind — CHANGES_REQUESTED** [review]
Can you rewrite this to not be Docker-based action. Docker-based actions rebuild the action container every time the workflow runs, which adds a non-insignificant time to how long the workflow takes. Please consider rewriting this as a [composite action](https://docs.github.com/en/free-pro-team@latest/actions/creating-actions/creating-a-composite-run-steps-action).

**Comment by jlapierre0istream** [comment]
> Can you rewrite this to not be Docker-based action. Docker-based actions rebuild the action container every time the workflow runs, which adds a non-insignificant time to how long the workflow takes. Please consider rewriting this as a [composite action](https://docs.github.com/en/free-pro-team@latest/actions/creating-actions/creating-a-composite-run-steps-action).

@marcind I refactored this to be a composite action.

File: get-cluster-info/src/main.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    for f in file_list:
+        clusters.append(Cluster(f))
+
+    pool = Pool(THREADPOOL_SIZE)
+    pool.map(runner, clusters)
+    pool.close()
+    pool.join()
+
+    md_data = create_markdown(clusters, TEMPLATE_DATA)
+    write_file(f"{BASE_PATH}/CLUSTERINFO.md", md_data)
```
Comment (CharlieSu) [inline get-cluster-info/src/main.py:90]: I'd prefer if the file name and path were configurable.

**Review by CharlieSu — CHANGES_REQUESTED** [review]

File: get-cluster-info/src/main.py (hunk 1 above)
Comment (jlapierre0istream) [reply get-cluster-info/src/main.py]: Fixed

File: get-cluster-info/src/main.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            with open(tools_file) as f:
+                match = re.match(HELM_VERSION_REGEX, f.read())
+            self.helm_version = match.group(1)
+        except Exception:
+            self.helm_version = "unknown"
+
+    def get_cluster_version(self):
+        try:
+            subprocess.run(["asdf", "install"], check=True, capture_output=True,
+                           cwd=self.cluster_path, text=True)
```
Comment (marcind) [inline get-cluster-info/src/main.py:48]: This assumes that the target uses asdf. This dependency should be removed. If a particular environment uses asdf, then asdf can be invoked explicitly before this action is executed.

File: get-cluster-info/src/main.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+{% for cluster in clusters -%}
+| {{cluster.name}} | {{cluster.cluster_version}} | {{cluster.region}} | {{cluster.helm_version}} |
+{% endfor %}
+"""
+
+
+class Cluster:
+    def __init__(self, cluster_path):
+        self.cluster_path = cluster_path
+        self.helm_path = (self.cluster_path).replace('clusters', 'helmfiles')
```
Comment (marcind) [inline get-cluster-info/src/main.py:32]: This rename captures an assumption that is very specific to a particular repository structure. We should remove this assumption, and if necessary allow the right parameters to be passed into the action.

One way to do that would be to require an action input parameter containing a glob pattern defining the set of directories containing cluster definitions.

File: get-cluster-info/src/main.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        self.name = os.path.basename(self.cluster_path)
+        self.region = os.path.basename(os.path.dirname(self.cluster_path))
+
+    def get_helm_version(self):
+        tools_file = f"{self.helm_path}/.tool-versions"
```
Comment (marcind) [inline get-cluster-info/src/main.py:37]: This also makes the assumption that `asdf` is being used. In addition, if we want to support helm2 (which we may, we still need to make a call to see what is actually running on the server, not just look at whatever version of the client happens to be active)

File: get-cluster-info/src/main.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+
+def runner(cluster):
+    cluster.get_helm_version()
+    cluster.get_cluster_version()
+
+
+def main():
+    clusters = []
+    file_list = sorted(generate_file_list(BASE_PATH, 'provider.tf'))
```
Comment (marcind) [inline get-cluster-info/src/main.py:87]: This assumes that a particular naming convention is followed, which may not always be the case.

Instead, the list of workspaces to inspect should be passed in via an action input variable (preferably in the form of a glob pattern)

**Review by marcind — CHANGES_REQUESTED** [review]
I think this implementation is too tied to the layout of a particular repository. I think we can improve on it to get something more broadly applicable.
If there's not a need to implement something more generic then this action can also be moved to the repository that that it's currently designed for.

**Comment by jlapierre0istream** [comment]
@marcind I'm going to take your suggestion and move this into the repository that it will be used in since it does not need to be very generic.

**Pull request closed without merging**